### PR TITLE
Rename InternedString to LocalInternedString and introduce a new thread-safe InternedString

### PR DESCRIPTION
**Pull request by Zoxc**
This is an allocation-free alternative to https://github.com/rust-lang/rust/pull/46972.

cc @rust-lang/compiler 

r? @michaelwoerister

### CODE REVIEW

**Comment by petrochenkov** [comment]
Why this distinction is necessary? When one should use `InternedString` and when `LocalInternedString`? I can't infer it from this patch (I haven't followed other parallelization PRs though).

I planned to get rid of `InternedString` completely in favor of `Symbol` after https://github.com/rust-lang/rust/issues/49300 is fixed (and `StableInternedString` -> `StableSymbol` for string-y `Ord` and stuff).

**Comment by Zoxc** [comment]
> Why this distinction is necessary?

`LocalInternedString` has a `Deref<Target=str>` impl which makes it more expressive to use, but also makes it not thread-safe (it doesn't implement `Send`). `InternedString` has no `Deref<Target=str>` impl and implements `Send` and `Sync` which makes it suitable to be shared between multiple threads.

> When one should use InternedString and when LocalInternedString?

Pick `InternedString` if you are sharing or sending things between threads (the compiler will inform you). Otherwise you can pick either one.

> I planned to get rid of InternedString completely in favor of Symbol

You can still get rid of the introduced `InternedString` in favor of Symbol. It still makes sense to keep around `LocalInternedString` because of its more expressive API, but it should probably be renamed to `LocalSymbol` at that point.

**Comment by petrochenkov** [comment]
I've just noticed that string interner is no longer thread local and every access through `Symbol`'s index into interner's vector goes through a lock.
So it looks like `LocalInternedString` in addition to being a convenience also serves as an optimization.
For convenience alone I'd personally wouldn't add an extra string type, but with locks I guess there's no way to avoid this string type proliferation. Sigh.

**Comment by Zoxc** [comment]
@petrochenkov We can replace everything with `Symbol(&'tcx str)` once querification reaches libsyntax =P

**Comment by petrochenkov** [comment]
>interner is no longer thread local and every access through Symbol's index into interner's vector goes through a lock

Actually, now I understand even less. If the interner is global, then string contents live effectively forever once interned and can be accessed from all threads, then why `Deref` to `str` is incompatible with `Sync`/`Send`. Please, ELI5 how this works.

**Comment by Mark-Simulacrum** [comment]
I'm also somewhat confused in general about the design of `Interner`; it feels like we could relatively easily get away with something like what is implemented in this [repo](https://github.com/Mark-Simulacrum/string-internment/blob/master/src/lib.rs) -- the API isn't one-to-one compatible, but the potential API should be effectively the same as today's Symbol/Interner API, if not easier to use. The approach is one where we basically just leak the data (which must be true for today's implementation of `Interner` if it's safe, since we have safe, public `as_str` on symbol that returns `&'static str`.

The negative there is that the size of `Symbol` increases from 4 bytes to 16 bytes as it goes from `u32` to `&'static str`. I don't know what the size and performance impact of that would be.

----

This might be best to discuss elsewhere, but it's also confusing to me why we make `InternedString` not implement `Send` when it does implement `Sync`. Since it's a `Copy` type that *seems* to imply that not implementing Send means nothing -- I can share a reference between two threads and then copy it in both threads. (Maybe this means a lint on this situation would be good?)

**Comment by Zoxc** [comment]
> The approach is one where we basically just leak the data (which must be true for today's implementation of Interner if it's safe, since we have safe, public as_str on symbol that returns &'static str.

We do not want to leak strings. The idea of `LocalInternedString` is that it contains a `&'thread str` where `thread` is the lifetime of the current thread. So in order to ensure `LocalInternedString` is sound we just need to have the interner (which contains the `str`) outlive threads we create `LocalInternedString`s on.

> This might be best to discuss elsewhere, but it's also confusing to me why we make InternedString not implement Send when it does implement Sync. Since it's a Copy type that seems to imply that not implementing Send means nothing -- I can share a reference between two threads and then copy it in both threads. (Maybe this means a lint on this situation would be good?)

`LocalInternedString` can't implement both `Sync` and `Copy`, so I added a commit to remove the Sync impl.

**Comment by Mark-Simulacrum** [comment]
> We do not want to leak strings.

Yet isn't that basically a requirement for today's implementation?

"This accesses the thread-local interner in order to get to the string." is the comment on `InternedString` which can safely be retrieved from `LocalInternedString`. I also see "This is safe because the interner never frees strings" in the code; that also implies that we're effectively leaking strings today while still paying the overhead for keeping `InternedString` not just pointing to the string data, and instead having to go through Symbol to get to the data. Am I missing something? What is the difference between `LocalInternedString` and `InternedString`?

It seems to me that long-term `Symbol` should become `&'static str` (perhaps wrapped in a newtype) and that should be feasible and is *almost* what we have today anyway.

**Comment by Zoxc** [comment]
> Yet isn't that basically a requirement for today's implementation?

No. The requirement is that the interner outlives any threads using `LocalInternedString`. Since `LocalInternedString` is not `Send` nor `Sync` and the `&'static str` field is private, there is no way for the string reference to escape the thread using it.

> I also see "This is safe because the interner never frees strings" in the code

It never frees strings until it is dropped, then all the strings are freed. I changed this comment to make it clearer.

> "This accesses the thread-local interner in order to get to the string." is the comment on InternedString which can safely be retrieved from LocalInternedString

I'm not sure what your point is there, but the comment is incorrect since we the interner is no longer thread-local. I removed that.

**Comment by Mark-Simulacrum** [comment]
Ah, I think I understand now. I believe the root of my confusion came from the fact that without parallel queries being enabled, neither `InternedString` nor `LocalInternedString` are `Sync/Send`, which makes the API this PR proposes fine.

In order for this to be fine with parallel queries, I believe the Interner must live globally (not per-thread); is that an accurate representation of the code today? Otherwise as far as I can tell a `Symbol` created on thread A can be transferred to other threads, which would lead to problems.

**Comment by Zoxc** [comment]
cc @rust-lang/compiler - Just wondering if there are any objections to this

**Comment by michaelwoerister** [comment]
@Mark-Simulacrum, would you mind starting a perf run for this?

Link: http://perf.rust-lang.org/compare.html?start=85f5dd489ea1cb93707ba1ddf4a7678d3aee3f74&end=7ef4d38de555741a0008ab09b945d77a2e9c0b65&stat=instructions%3Au

**Comment by Mark-Simulacrum** [comment]
Perf queued.

**Comment by michaelwoerister** [comment]
OK, performance looks good! r=me after rebase.

File: src/libsyntax_pos/symbol.rs [hunk 1]
```diff
@@ -365,84 +377,208 @@ fn with_interner<T, F: FnOnce(&mut Interner) -> T>(f: F) -> T {
...
+impl Encodable for LocalInternedString {
+    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        s.emit_str(self.string)
+    }
+}
+
+/// Represents a string stored in the string interner
+#[derive(Clone, Copy, Eq)]
+pub struct InternedString {
+    symbol: Symbol,
```
Comment (eddyb) [inline src/libsyntax_pos/symbol.rs]: I saw this PR too late, but wish this was named `SymbolContents` instead of `InternedString`, since AFAICT all it does is it changes the behavior of comparison / hashing / etc. to act on the string value inside.

Comment (eddyb) [reply src/libsyntax_pos/symbol.rs]: `SymbolAsStr`? I mean, I want the type to evoke "this is a `Symbol` made to look like the string contents thereof".
